stage1/tests/test_upgrade_ordering.py

```
@@ -0,0 +1,708 @@
+    """A worker must not drain itself unless the control-plane play signed off.
+
+    Without the cross-host gate a worker would drain against a control plane whose preflight
+    never ran, or that failed its own upgrade.
+    """
+    gate = None
+    for task, ancestors in walk(load_tasks(AGENT_UPGRADE)):
+        if (task.get("vars") or {}).get(STEP_VAR) == "drain":
+            gate = gate_text(task, ancestors)
+            break
+    if gate is None:
+        raise HarnessError(f"no drain-marked task in {rel(AGENT_UPGRADE)}")
```

Ruff (0.16.0)

[warning] Avoid specifying long messages outside the exception class

(TRY003)